src/python/pants/reporting/async_workunit_handler.py

```diff
@@ -0,0 +1,40 @@
+# Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+import threading
+from typing import Any, Callable
+
+
+class AsyncWorkunitHandler:
+  def __init__(self, scheduler: Any, callback: Callable, report_interval_seconds: float = 0.01):
```

Comment:
100 times per seconds seems very frequent as a default. I can see the reporting having a pretty large fixed cost (like millisecond - many milliseconds). I would probably be more comfortable with 10 times per second as a guess, or a comment backing up the decision with some data that shows it won't be too expensive.

Reply:
As far as I know, this will acquire the GIL every time, so doing it too frequently actually has cost. I agree that if we want to do this in python it should probably be much less frequent (1/second maybe?); alternatively, Rust wouldn't have the GIL issue.

Reply:
Absolutely - I set this value low because I was testing this locally on some quick targets and I wanted to make sure I saw some output before it ended. Naming the magic number is a good idea, and I was imagining that the "actual" default for this would be something like 10s of seconds (since our use case is making periodic web requests).

Reply:
@illicitonion agreed, acquiring the GIL too frequently would make this strategy fruitless. I went with implementing the threading in python because I had trouble figuring out where `InnerSession` finally got dropped (in a rust context), and because we expect to invoke this with a low enough frequency (10s of seconds) that too-frequent GIL locking isn't an issue. But if I could figure out where it's okay to start/stop the thread code in rust, I'd be happy to do it there and avoid the issue.

Reply:
Cool, as long as the frequency is different I don't think this is too scary :)

I don't think it would be the end of the world, also, to just have a thread started on `Core` and have the queue be a process-global thing rather than per-`Session`. Possibly that means that we could even avoid re-connection overhead for repeated invocations with a shared daemon.

But I think this is fine as-is too :)